validator/beacon/service.go

```diff
@@ -61,68 +87,95 @@ func (s *Service) Stop() error {
 	}
-	for {
-		block, err := stream.Recv()
+}
 
-		// If the stream is closed, we stop the loop.
-		if err == io.EOF {
-			break
-		}
-		if err != nil {
-			log.Errorf("Could not receive latest beacon block from stream: %v", err)
-			continue
-		}
-		log.WithField("slotNumber", block.GetSlotNumber()).Info("Latest beacon block slot number")
-
-		// Based on the slot determined from the latest crystallized state, check if
-		// it matches the latest received beacon slot.
-		if s.responsibility == "proposer" {
-			log.WithField("slotNumber", block.GetSlotNumber()).Info("Assigned proposal slot number reached")
-			s.responsibility = ""
-			s.proposerAssignmentFeed.Send(block)
-		} else if s.responsibility == "attester" && block.GetSlotNumber() == s.assignedSlot {
-			// TODO: Let the validator know a few slots in advance if its attestation slot is coming up
-			log.Info("Assigned attestation slot number reached")
-			s.responsibility = ""
-			s.attesterAssignmentFeed.Send(block)
+// waitForAssignment kicks off once the validator determines the currentSlot of the
+// beacon node by calculating the difference between the current system time
+// and the genesis timestamp. It runs exactly every SLOT_LENGTH seconds
+// and checks if it is time for the validator to act as a proposer or attester.
+func (s *Service) waitForAssignment(ticker <-chan time.Time, client pb.BeaconServiceClient) {
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case <-ticker:
+			log.WithField("slotNumber", s.currentSlot).Info("New beacon node slot interval")
+			if s.responsibility == "proposer" && s.assignedSlot == s.currentSlot {
+				log.WithField("slotNumber", s.currentSlot).Info("Assigned proposal slot number reached")
+				s.responsibility = ""
+				block, err := client.CanonicalHead(s.ctx, &empty.Empty{})
+				if err != nil {
+					log.Errorf("Could not fetch canonical head via gRPC from beacon node: %v", err)
+					continue
+				}
+				// We forward the latest canonical block to the proposer service via a feed.
+				s.proposerAssignmentFeed.Send(block)
+			} else if s.responsibility == "attester" && s.assignedSlot == s.currentSlot {
+				log.Info("Assigned attestation slot number reached")
+				s.responsibility = ""
+				block, err := client.CanonicalHead(s.ctx, &empty.Empty{})
+				if err != nil {
+					log.Errorf("Could not fetch canonical head via gRPC from beacon node: %v", err)
+					continue
+				}
+				// We forward the latest canonical block to the attester service a feed.
+				s.attesterAssignmentFeed.Send(block)
+			}
+			// Increase the current slot by one every 8 seconds.
+			s.currentSlot++
 		}
 	}
 }
 
-func (s *Service) fetchCrystallizedState(client pb.BeaconServiceClient) {
-	var activeValidatorIndices []int
+// listenForCrystallizedStates receives the latest canonical crystallized state
+// from the beacon node's RPC server via gRPC streams.
+// TODO: Rename to listen for assignment instead, which is streamed from a beacon node
```

Comment:
Open up an issue for this , or reference this to another issue

Reply:
+1 issue formatting of TODO